extensions/kubernetes/spi/src/main/java/io/quarkus/kubernetes/spi/KubernetesEnvBuildItem.java

```diff
@@ -5,26 +5,96 @@
 
 public final class KubernetesEnvBuildItem extends MultiBuildItem {
 
+    public static ItemKey keyFor(EnvType type, String name) {
+        return new ItemKey(type, name);
+    }
+
+    public static ItemKey keyFor(KubernetesEnvBuildItem item) {
+        return new ItemKey(item.getType(), item.getName());
+    }
+
+    public enum EnvType {
+        var(false),
+        field(false),
+        secret(true),
+        configmap(true);
+
+        public final boolean allowMultipleDefinitions;
+
+        EnvType(boolean allowMultipleDefinitions) {
+            this.allowMultipleDefinitions = allowMultipleDefinitions;
+        }
+
+        public boolean isCompatibleWith(EnvType type) {
+            if (this == type) {
+                return true;
+            }
+
+            switch (this) {
+                case field:
+                    return type == var;
+                case var:
+                    return type == field;
+                case secret:
+                    return type == configmap;
+                case configmap:
+                    return type == secret;
+                default:
+                    return false;
+            }
+        }
+    }
+
     private final String name;
     private final String value;
-    private final String secret;
-    private final String configmap;
-    private final String field;
+    private final EnvType type;
     private final String target;
 
     public KubernetesEnvBuildItem(String name, String value, String target) {
-        this(name, value, null, null, null, target);
+        this(name, value, EnvType.var, target);
     }
 
-    public KubernetesEnvBuildItem(String name, String value, String secret, String configmap, String field, String target) {
+    public static EnvType getEnvType(String secret, String configmap, String field) {
+        final EnvType type;
+        if (secret != null) {
+            type = EnvType.field;
+        } else if (configmap != null) {
+            type = EnvType.configmap;
+        } else if (field != null) {
+            type = EnvType.field;
+        } else {
+            type = EnvType.var;
+        }
+        return type;
+    }
+
+    public KubernetesEnvBuildItem(String name, String value, EnvType type, String target) {
```

Comment:
This is me being compulsive, but since the `key` is (key, name), how about we bring the type next to the name?
```suggestion
    public KubernetesEnvBuildItem( EnvType type, String name, String value, String target) {
```